### ISSUE TITLE
Bug with azurerm_monitor_diagnostic_setting and dynamic inline blocks

### ISSUE THREAD

**Issue by nnellans-acts** [terraform-providers/terraform-provider-azurerm#9923]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.14.2
AzureRM Provider 2.41.0

### Affected Resource(s)

* `resource azurerm_monitor_diagnostic_setting`
* `data azurerm_monitor_diagnostic_categories`

### Terraform Configuration Files

```hcl
data "azurerm_monitor_diagnostic_categories" "example" {
  resource_id = azurerm_virtual_desktop_application_group.example.id
}

resource "azurerm_monitor_diagnostic_setting" "example" {
  name = "log_all"
  target_resource_id = azurerm_virtual_desktop_application_group.example.id
  log_analytics_workspace_id = azurerm_log_analytics_workspace.example.id
		
  dynamic "log" {
    for_each = data.azurerm_monitor_diagnostic_categories.example.logs

    content {
      category = log.value
    }
  }
		
  dynamic "metric" {
    for_each = data.azurerm_monitor_diagnostic_categories.example.metrics

    content {
      category = metric.value
    }
  }
}
```

### Expected Behaviour

All Diagnostic Settings enabled for the resource

### Actual Behaviour

```hcl
Error: Provider produced inconsistent final plan

When expanding the plan for azurerm_monitor_diagnostic_setting.example
to include new values learned so far during apply, provider
"registry.terraform.io/hashicorp/azurerm" produced an invalid new value for
.log: block set length changed from 1 to 3.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.


Error: Provider produced inconsistent final plan

When expanding the plan for azurerm_monitor_diagnostic_setting.example
to include new values learned so far during apply, provider
"registry.terraform.io/hashicorp/azurerm" produced an invalid new value for
.metric: planned set element
cty.ObjectVal(map[string]cty.Value{"category":cty.UnknownVal(cty.String),
"enabled":cty.True,
"retention_policy":cty.ListValEmpty(cty.Object(map[string]cty.Type{"days":cty.Number,
"enabled":cty.Bool}))}) does not correlate with any element in actual.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Steps to Reproduce

1. `terraform apply`

**Comment by rjfmachado** [comment]
I can reproduce, my scenario is azurerm_network_security_group failing. AKS and other resources work well.

```terraform
terraform {
  required_version = ">= 0.12"
  required_providers {
    azurerm = ">= 2.0"
    random  = ">= 2.0"
  }
}

provider "azurerm" {
  features {
    virtual_machine {
      delete_os_disk_on_deletion = true
    }
  }
}

provider "random" {}

resource "random_pet" "nsg" {
  length    = 2
  prefix    = "nsg"
  separator = ""
}

resource "azurerm_resource_group" "rg" {
  name     = random_pet.nsg.id
  location = "West Europe"
}

resource "azurerm_network_security_group" "nsg" {
  name                = "bastion"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
}

resource "azurerm_log_analytics_workspace" "logs" {
  name                = random_pet.nsg.id
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  sku                 = "PerGB2018"
}

data "azurerm_monitor_diagnostic_categories" "nsgdiags" {
  resource_id = azurerm_network_security_group.nsg.id
}

resource "azurerm_monitor_diagnostic_setting" "diags" {
  name                       = "nsg"
  target_resource_id         = azurerm_network_security_group.nsg.id
  log_analytics_workspace_id = azurerm_log_analytics_workspace.logs.id

  dynamic "log" {
    for_each = data.azurerm_monitor_diagnostic_categories.nsgdiags.logs
    content {
      category = log.value
      enabled  = true

      retention_policy {
        enabled = false
      }
    }
  }

  dynamic "metric" {
    for_each = data.azurerm_monitor_diagnostic_categories.nsgdiags.metrics
    content {
      category = metric.value

      retention_policy {
        enabled = false
      }
    }
  }
}
```

**Comment by magodo** [comment]
hashicorp/terraform#28340 is now fixed in Terraform v0.15.2, on which version this issue should now be addressed.

**Comment by katbyte** [comment]
closing as @magodo indicated this has been fixed a number of months ago
